addons/udes_stock/models/stock_picking.py

```diff
@@ -1769,14 +1769,21 @@ def reserve_stock(self):
                 pickings = moves.mapped('picking_id')
                 processed |= pickings
 
-                if not picking_type.u_handle_partials:
-                    unsatisfied = pickings.filtered(
-                        lambda x: x.state != 'assigned')
-                    if unsatisfied:
+                unsatisfied = pickings.filtered(
+                    lambda x: x.state != 'assigned')
+                mls = pickings.mapped('move_line_ids')
+                if unsatisfied:
+                    # Rollback if the picking type cannot handle partials or it
+                    # can but there is nothing allocated (no stock.move.lines)
+                    if not picking_type.u_handle_partials or not mls:
                         if self:
                             # we need to construct our error message before the
-                            # changes are rolled back.
-                            moves = unsatisfied.mapped('move_lines')
+                            # changes are rolled back, and report only products
+                            # that are unreservable.
+                            not_done = lambda x: x.state not in (
```

Comment:
Nitpick: there are a number of definitions inside the nested `for` `while` that can be moved outside; I suggest to address in a separate PR in the future.